### ISSUE TITLE
Single token refresh when restore N browser tabs

### ISSUE THREAD

**Issue by peluprvi** [websanova/vue-auth#298]
I'm studing Vue.js and I'm about to move to it (I don't know much for now). I subscribed to https://github.com/websanova/vue-auth/issues/289 as I had a similar issue on AngularJS with another library.

By @jap1968 description, vue-auth has a fixed 30 min between refresh requests and I also saw it has a refresh check on the loading process. This can trigger a 401 if the user restore N browser tabs and/or the library send two refresh requests with the same token at the same time.
That was my issue on AngularJS. The first request updates the token and invalidates the second request.

Steps to reproduce:
- login (with rememberMe: true) and create a valid token
- open some browser tabs (10 or more can be enough)
- close the browser (all tabs together)
- reopen the browser
- restore the last tabs in the moment the token needs refresh but still valid

@jap1968 I know it's not the same issue, but please can you check if this happens in your test setup?

@websanova Does vue-auth already prevent this?

**Comment by websanova** [comment]
There should be some grace period for expired tokens on the server.

**Comment by peluprvi** [comment]
Thank you for your return!

So, let me know if this will work:
- Following the steps above, you will get "Token A" before restore the tabs
- Token is about to expire/refresh
- After restore all tabs, the Tab 1 will refresh Token A and it gets Token B without revoke Token A
- Parallel to this, Tab 2 will also refresh Token A, but it gets Token C without revoke Token A or B

As you said, at this point, Token A, B and C are all valid. So, if any Tab send a request to the server, with Token A, B or C, will all requests return with success? And are they (tokens) all valid until the expiration date?

If this is right, my current process (on the server) is revoking the token somehow and I will need to debug it even before change move to Vue.js.

**Comment by websanova** [comment]
Yea, I think you got the right idea.

But it's not an issue with Vue.js. It's a known server side issue if you
are using tokens. Particularly with JWT.

You can of course have some check on local/browser side to try to repeat a
request, but to me that feels a bit messy. Not to mention if you have
browser and mobile apps.

I think the better way to handle it is on server. There should already be a
package for it with whatever language/framework you are using. But
basically it's just giving expired tokens another minute or so before they
are fully invalidated.

**Comment by jap1968** [comment]
@websanova said:
> There should be some grace period for expired tokens on the server.

Well, in fact there is some debate about that:
[Is refreshing an expired JWT token a good strategy?](https://security.stackexchange.com/questions/119371/is-refreshing-an-expired-jwt-token-a-good-strategy)

Our API implementation does not give such grace period. Tokens have to be refreshed before expiration. Otherwise our system returns a 401 code (and no refresh token at all).

**Comment by websanova** [comment]
ok,

well that's at the app level then really.

If one of the requests returns a 401 then I don't see anything you can do
other than try to make the request again (hopefully with a valid token on
the next request).

Personally, I don't bother much with refreshes, I just have a token that
lasts 2 weeks, and attempt to keep bumping it up by either fetching a new
token on refresh or on some arbitrary interval like 30 minutes.

If the user doesn't use the app for 2 weeks the token is expired and the
user must login again.

But I guess depends on the security level you need.

**Comment by peluprvi** [comment]
Thanks a lot to help me figure out that!

I found a not documented variable in our server vendor library and I changed it to not revoke the old token. It will allow us to simplify the entire client process.

About the expiration time and process, it's really depends on the application.

I'm closing this issue as vue-auth library allows this behavior to be implemented.
